source/opt/debug_info_manager.cpp

```diff
@@ -0,0 +1,198 @@
+      context(), SpvOpExtInst, context()->get_type_mgr()->GetVoidTypeId(),
+      result_id,
+      {
+          {SPV_OPERAND_TYPE_RESULT_ID,
+           {context()
+                ->get_feature_mgr()
+                ->GetExtInstImportId_OpenCL100DebugInfo()}},
+          {SPV_OPERAND_TYPE_EXTENSION_INSTRUCTION_NUMBER,
+           {static_cast<uint32_t>(OpenCLDebugInfo100DebugInfoNone)}},
+      }));
+
+  id_to_dbg_inst_[result_id] = dbg_info_none_inst.get();
+
+  // Add to the front of |ext_inst_debuginfo_|.
+  return context()->module()->ext_inst_debuginfo_begin()->InsertBefore(
+      std::move(dbg_info_none_inst));
+}
+
+Instruction* DebugInfoManager::CloneDebugDeclare(uint32_t from, uint32_t to) {
+  auto debugdecl_it = local_var_id_to_dbgdecl_.find(from);
+  if (debugdecl_it == local_var_id_to_dbgdecl_.end()) return nullptr;
+
+  auto* clone = debugdecl_it->second->Clone(context());
+  clone->SetResultId(context()->TakeNextId());
+  clone->GetOperand(kDebugDeclareOperandVariableIndex).words[0] = to;
+
+  if (local_var_id_to_dbgdecl_.find(to) == local_var_id_to_dbgdecl_.end())
+    local_var_id_to_dbgdecl_[to] = clone;
+  return clone;
+}
+
+Instruction* DebugInfoManager::GetDebugInlinedAt(uint32_t dbg_inlined_at_id) {
+  auto it_inlined_at = id_to_dbg_inst_.find(dbg_inlined_at_id);
+  if (it_inlined_at == id_to_dbg_inst_.end()) return nullptr;
+  if (it_inlined_at->second->GetOpenCL100DebugOpcode() !=
+      OpenCLDebugInfo100DebugInlinedAt) {
+    return nullptr;
+  }
+  return it_inlined_at->second;
+}
+
+Instruction* DebugInfoManager::CloneDebugInlinedAt(uint32_t dbg_inlined_at_id) {
+  auto* inlined_at = GetDebugInlinedAt(dbg_inlined_at_id);
+  if (inlined_at == nullptr) return nullptr;
+  auto* new_inlined_at = inlined_at->Clone(context());
+  new_inlined_at->SetResultId(context()->TakeNextId());
+  id_to_dbg_inst_[new_inlined_at->result_id()] = new_inlined_at;
+  return new_inlined_at;
+}
+
+void DebugInfoManager::AnalyzeDebugInsts(Module& module) {
+  for (auto& inst : module.ext_inst_debuginfo()) {
+    id_to_dbg_inst_[inst.result_id()] = &inst;
+    if (inst.GetOpenCL100DebugOpcode() == OpenCLDebugInfo100DebugFunction) {
+      auto fn_id =
+          inst.GetSingleWordOperand(kDebugFunctionOperandFunctionIndex);
+      assert(fn_id_to_dbg_fn_.find(fn_id) == fn_id_to_dbg_fn_.end() &&
+             "Two DebugFunction instruction exists for a single OpFunction.");
+      fn_id_to_dbg_fn_[fn_id] = &inst;
+    }
+  }
+
+  module.ForEachInst([this](Instruction* cpi) {
+    if (cpi->GetOpenCL100DebugOpcode() == OpenCLDebugInfo100DebugDeclare ||
+        cpi->GetOpenCL100DebugOpcode() == OpenCLDebugInfo100DebugValue) {
+      auto var_id =
+          cpi->GetSingleWordOperand(kDebugDeclareOperandVariableIndex);
+      if (local_var_id_to_dbgdecl_.find(var_id) !=
+          local_var_id_to_dbgdecl_.end()) {
+        return;
+      }
+      local_var_id_to_dbgdecl_[var_id] = cpi;
+    }
+  });
+}
```

Comment:
You will need a way for the users of the analysis to update it.  When the ssa-rewriter creates new debug instruction, the analysis may have to be updated.  What kind of instructions will inlining create?  It could create new debug scope instructions, but the analysis won't reflect this change.

I would suggest something like the def-use manager.  Write one function that handles a single instruction.  Make that public.  Then have the version that looks at the entire module call the single instruction version on every instruction that matters.

Reply:
I still want to see what I suggested above.  Have one function called `AnalyzeDebugInst(Intruction* dbg_inst)` that is called by this function.  This is a single generic function that is needed for inlining.  How will inlining handle a DebugValue instruction that needs to be inlined?